.github/workflows/pull-request-with-code.yml

```diff
@@ -17,35 +17,62 @@ on:
       - '**/*.toml'
   workflow_dispatch:
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/master' }}
+
 env:
-  ORG_GRADLE_PROJECT_signingKey : ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGKEY }}
+  ORG_GRADLE_PROJECT_signingKey: ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGKEY }}
   ORG_GRADLE_PROJECT_signingKeyId: ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGKEYID }}
-  ORG_GRADLE_PROJECT_signingKeyPassword : ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGPASSWORD }}
+  ORG_GRADLE_PROJECT_signingKeyPassword: ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGPASSWORD }}
   CLI_TEST_MAX_DURATION_IN_SECONDS: 10
 
 # Note that "jobs.build.strategy" and "jobs.build.runs-on" should be kept in sync with "pull-request-wth-code"
 jobs:
   build:
     strategy:
       matrix:
-        os: [ ubuntu-latest, windows-latest ]
-        # When changing the list of JDK versions, the build configuration has to be changed by a repository admin. See
-        # https://github.com/pinterest/ktlint/pull/1787#issuecomment-1409074092
-        jdk: [ 8, 11, 17, 19 ]
-        exclude: # windows with JDK8 are *really* flaky
-          - os: windows-latest
-            jdk: 8
+        os: [ubuntu-latest, windows-latest]
     runs-on: ${{ matrix.os }}
+    name: "[build] OS: ${{ matrix.os }}"
     steps:
       - uses: actions/checkout@v3
+
       - uses: actions/setup-java@v3
         with:
           distribution: 'zulu'
-          java-version: ${{ matrix.jdk }}
+          java-version: 20 # This should match all `jvmToolchain(xxx)` calls in the project
+
       - uses: gradle/gradle-build-action@v2
         with:
           gradle-home-cache-cleanup: true
+
       - name: Build with release Kotlin version
         run: ./gradlew build ktlintCheck --no-configuration-cache
-      - name: Build with dev Kotlin version
+
+      - name: Build with assemble Kotlin version
         run: ./gradlew -PkotlinDev build ktlintCheck --no-configuration-cache
+
+  tests:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, windows-latest]
+        jdk: [ 8, 11, 17 ] # list of Java versions to run tests against
```

Comment:
This doesn't have to be kept in sync with buildscript anymore. This is the only place where the tested configuration are defined. If one wants to start testing on Java 27 - they can just add another argument
```kotlin
        jdk: [ 8, 11, 17, 27 ]
```

`actions/setup-java` will install proper Java version, `testOnJdk -PtestJdkVersion=27` will invoke tests with proper config

Reply:
Also, a small note: the tests against _current_ version are invoked as a part of `build` job. If the main pipeline takes too much time, you may consider running them on a separate node, in parallel to the _main_ `build` job. 
This can be done by using `./gradlew build -x test` + adding current java major version to the `jdk` list here.

Reply:
Please document that current java-version does not need to be listed here.
```
jdk: [ 8, 11, 17 ] # list of Java versions (excluding java-compatibility version for which tests already have run during build)  to run tests against
```